src/transformers/models/modernbert/modular_modernbert.py

```diff
@@ -0,0 +1,1515 @@
+        raise RuntimeError("When using the full megatron init, every module must have a type.")
+
+    cutoff_factor = config.initializer_cutoff_factor
+    if cutoff_factor is None:
+        cutoff_factor = 3
+
+    if module_type == ModernBertModuleType.in_module:
+        std = config.initializer_range  # for att_proj (same as QKV), ff_proj
+    elif module_type == ModernBertModuleType.out_module:
+        std = config.initializer_range / math.sqrt(2.0 * config.num_hidden_layers)  # for attn_out, ff_out
+    elif module_type == ModernBertModuleType.embedding:
+        std = config.initializer_range  # token embeddings (wte)
+    elif module_type == ModernBertModuleType.final_out:
+        std = config.hidden_size**-0.5  # final output (ff_out)
+    else:
+        raise RuntimeError(f"Unknown module type '{module_type}'")
+
+    nn.init.trunc_normal_(
+        module.weight,
+        mean=0.0,
+        std=std,
+        a=-cutoff_factor * std,
+        b=cutoff_factor * std,
+    )
+
+    if isinstance(module, nn.Linear):
+        if module.bias is not None:
+            nn.init.zeros_(module.bias)
+
+
+def _unpad_modernbert_input(
+    inputs: torch.Tensor,
+    attention_mask: torch.Tensor,
+    position_ids: Optional[torch.Tensor] = None,
+    labels: Optional[torch.Tensor] = None,
+) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor, int, Optional[torch.Tensor], Optional[torch.Tensor]]:
+    """
+    Remove padding from input sequences.
+
+    Args:
+        inputs: (batch, seqlen, ...) or (batch, seqlen)
+        attention_mask: (batch, seqlen), bool / int, 1 means valid and 0 means not valid.
+        position_ids: (batch, seqlen), int, position ids
+        labels: (batch, seqlen), int, labels
+
+    Returns:
+        unpadded_inputs: (total_nnz, ...), where total_nnz = number of tokens selected in attention_mask.
+        indices: (total_nnz)
+        cu_seqlens: (batch + 1), the cumulative sequence lengths
+        max_seqlen_in_batch: int
+        unpadded_position_ids: (total_nnz) or None
+        unpadded_labels: (total_nnz) or None
+    """
+    seqlens_in_batch = attention_mask.sum(dim=-1, dtype=torch.int32)
+    indices = torch.nonzero(attention_mask.flatten(), as_tuple=False).flatten()
+    max_seqlen_in_batch = int(seqlens_in_batch.max().item())
+    cu_seqlens = torch.nn.functional.pad(torch.cumsum(seqlens_in_batch, dim=0, dtype=torch.int32), (1, 0))
+
+    if inputs.dim() == 2:
+        unpadded_inputs = inputs.flatten()[indices]
+    else:
+        batch, seqlen, *rest = inputs.shape
+        shape = batch * seqlen
+        unpadded_inputs = inputs.view(shape, *rest)[indices]
+
+    unpadded_position_ids = position_ids.flatten()[indices] if position_ids is not None else None
+    unpadded_labels = labels.flatten()[indices] if labels is not None else None
+
+    return unpadded_inputs, indices, cu_seqlens, max_seqlen_in_batch, unpadded_position_ids, unpadded_labels
+
+
+def _pad_modernbert_output(
```

Comment:
we actually also have utils for that here: https://github.com/huggingface/transformers/blob/1b6cb1eefc140ea2dbd74e23389050ebcb7f417d/src/transformers/modeling_flash_attention_utils.py#L310-L324

Reply:
Ideally we'll avoid `flash_attn` their `pad_input` for the repadding, so that we can also support unpadded FlexAttention without the `flash_attn` dependency.
Currently `flash_attn` is still needed because of the `RotaryEmbedding` from `flash_attn`, but we're interested in a torch-only version as a fallback.

I will shrink `_pad_modernbert_output` though: we're not using the `padded_labels` anywhere.

Reply:
Shrunk/removed unused path in b241a7e97a1413be7efcbf0e1435dd1581a2dc0f.